### PR TITLE
Master unify top bar and tool bar

### PR DESCRIPTION
**Pull request by laa-odoo**
## Description:

Intro: Currently, the topbar and the toolbar share common functionalities: The majority of shortcuts present in the toolbar are present in the topbar. Also, the toolbar must be able to call the same menus present in the topbar (cf, number format, font size, formulas)

Problem: Today, all the code that governs the toolbar is a separate code. So all the common functionalities with the topbar are developed twice, which isn't great to maintain... Also, the toolbar calls its own menus, which are visually different from the topbar menus...

Solution: to answer the problem, this pr proposes to factorize all the common functionalities at the scale of the menu items. Thus, the menu items can always be called in the menu registries to compose the topbar, but they can also be used in the toolbar in the form of a button.

This technical factorization will require small functional changes: indeed, currently, the menus of the toolbar have something more than the menus of the topbar: they have a state resuming the current state of the selected cells. For example, the number format menu of the toolbar has notch icons to indicate which format is selected.

Having the same behavior on topbar menus would involve using icons on menu items. This pr therefore proposes to generalize the use of icon on menu items. It is currently already possible to use icons on item menus, but their use is not great: icons conflict with the description of menu items.

So:
- The first commits of this PR will aim to improve the use of icons on menu items and will generalize the use of icon on all menus

- Latest commits will factor common functionalities across menu items and replace toolbar code with buttons using those menu items

Odoo task ID : [3145722](https://www.odoo.com/web#id=3145722&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/font_size_editor/font_size_editor.ts [hunk 1]
```diff
@@ -31,6 +31,14 @@ css/* scss */ `
     input::-webkit-inner-spin-button {
       -webkit-appearance: none;
     }
+
```
Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: menu-item height in context menus went down, 28px -> 26px

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: menu-item horizontal padding went down 16px -> 13px

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: Probably already there in master, but colorPicker dropdown is lower than all the other menus in the topbar

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: you forget to change the [selected] icon in menus + it's not centered 
image

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: nitpick : might need more shadows or something in the horitonzal menus of the toolbar, it kinda melds with the bottom border 
image

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: Maybe a bit less padding for menu w/ selected items ? The white space looks huge with regards to the menu width.
I think either the menu should be bigger or the blank space smaller
image

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: filter icon not centered in data menu 
image

Comment (hokolomopo) [inline src/components/font_size_editor/font_size_editor.ts:34]: Most likely nothing to do with your changes but while we're at it: 
font selector dropdown items don't have `cursor : pointer` + have some horizontal margin visible on hover contrarily to all the other menus.

**Review by hokolomopo — COMMENTED** [review]
Some comments on the style w/o reading the code at all.
Idk which changes were intentional/PO validated or just random

File: src/components/menu/menu.xml [hunk 2]
```diff
@@ -19,17 +19,22 @@
...
-            <span class="o-menu-item-name" t-esc="getName(menuItem)"/>
-            <span
-              t-if="menuItem.description"
-              class="o-menu-item-description"
-              t-esc="menuItem.description"
-            />
+            <span>
+              <span t-if="isIconMenu" class="o-menu-item-icon">
+                <t t-set="iconName" t-value="getIconName(menuItem)"/>
+                <t t-if="iconName" t-call="{{iconName}}"/>
```
Comment (hokolomopo) [inline src/components/menu/menu.xml:25]: is the `t-set` really needed here ?

File: src/components/menu_item_button/menu_item_button.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+const FONT_COLOR = "#4a4f59";
+const ACTIVE_BG_COLOR = "#E6F4EA";
+const ACTIVE_FG_COLOR = "#188038";
+const HOVERED_BG_COLOR = "#f6f7fa";
+const HOVERED_FG_COLOR = "#000000";
+
+//------------------------------------------------------------------------------
+
+css/* scss */ `
+  .o-tool {
```
Comment (hokolomopo) [inline src/components/menu_item_button/menu_item_button.ts:15]: We probably should have a better class name than `tool`, it doesn't make that much sense in a `MenuButton` component

File: src/components/menu_item_button/menu_item_button.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+export class MenuItemButton extends Component<Props, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-MenuItemButton";
+
+  private menuItem = createMenuItem(this.props.menuItemSpec);
+
+  get isVisible() {
+    return this.menuItem.isVisible(this.env);
+  }
```
Comment (hokolomopo) [inline src/components/menu_item_button/menu_item_button.ts:65]: unused

File: src/components/menu_item_button/menu_item_button.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  get isEnable() {
```
Comment (hokolomopo) [inline src/components/menu_item_button/menu_item_button.ts:67]:
```suggestion
  get isEnabled() {
```

File: src/components/font_size_editor/font_size_editor.ts (hunk 1 above)
Comment (laa-odoo) [reply src/components/font_size_editor/font_size_editor.ts:34]: yes, still some work to do on the size and dimension of the SVGs

Comment (hokolomopo) [reply src/components/font_size_editor/font_size_editor.ts:34]: I agree, but the menu should probably have a dynamic width for that, (there is a task somewhere), to not have a menu 200px wide to display a number

File: src/registries/menu_items_registry.ts [hunk 6]
```diff
@@ -32,30 +32,32 @@ export interface MenuItemSpec {
...
 export interface MenuItem {
-  id: string;
-  sequence: number;
   name: (env: SpreadsheetChildEnv) => string;
+  description: string;
+  sequence: number;
+  id: string;
   isVisible: (env: SpreadsheetChildEnv) => boolean;
   isEnabled: (env: SpreadsheetChildEnv) => boolean;
+  isSelected?: (env: SpreadsheetChildEnv) => boolean;
```
Comment (hokolomopo) [inline src/registries/menu_items_registry.ts:51]: Maybe make it mandatory, with a default value `() => false` to make it easier to use/consistent with isVisible/isEnabled

File: src/registries/menus/items/format_menu_items.ts [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+  name: _lt("Top"),
+  action: (env: SpreadsheetChildEnv) => ACTIONS.setStyle(env, { verticalAlign: "top" }),
+  isSelected: (env) => env.model.getters.getCurrentStyle().verticalAlign === "top",
+  icon: "o-spreadsheet-Icon.ALIGN_TOP",
+};
+
+export const formatAlignmentMiddleMenuItem: MenuItemSpec = {
+  name: _lt("Middle"),
+  action: (env: SpreadsheetChildEnv) => ACTIONS.setStyle(env, { verticalAlign: "middle" }),
+  isSelected: (env) => (env.model.getters.getCurrentStyle().verticalAlign || "middle") === "middle",
```
Comment (hokolomopo) [inline src/registries/menus/items/format_menu_items.ts:186]: We should have a constant `DEFAULT_VERTICAL_ALIGN` and use it here and in the renderer. Smae for the wrapping.

File: src/registries/menus/items/sheet_menu_itemps.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
+import { buildSheetLink, markdownLink } from "../../../helpers";
```
Comment (hokolomopo) [inline src/registries/menus/items/sheet_menu_itemps.ts:1]: typo in file name

File: src/registries/menus/items/sheet_menu_itemps.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import { _lt } from "../../../translation";
+import { MenuItemSpec } from "../../menu_items_registry";
+
+export const linkSheetMenuItem: MenuItemSpec = {
```
Comment (hokolomopo) [inline src/registries/menus/items/sheet_menu_itemps.ts:5]: Not sure it's worth to separate into 2 files `MenuItems` and `MenuRegistry` for the menu items that are only used once. It's a bit more annoying to use.

**Review by hokolomopo — COMMENTED** [review]
I really like it ! Makes the toolbar way better 👍

File: src/registries/menu_items_registry.ts (hunk 6 above)
Comment (laa-odoo) [reply src/registries/menu_items_registry.ts:51]: Yes and no:
have it optional, allow knowing if we shoud keep a space to accommodate the "check icon".

For example, if number formats of a selected cell does not correspond to the format offered in number format menu:
--> no format will be selected in the menu, 
--> but we want to leave empty spaces to make the user understand that the current format not correspond 

Conversely, there are menus that will never have the opportunity to have menu items selected or with default icons. In these cases, it is useless to have a space

See here: https://github.com/odoo/o-spreadsheet/pull/2046/files#diff-5a985db40b62ab2036f2e38176aaeb51df0a0da1ceef433f81b698d0ba783b6cR164

File: src/registries/menus/items/sheet_menu_itemps.ts (hunk 9 above)
Comment (laa-odoo) [reply src/registries/menus/items/sheet_menu_itemps.ts:5]: Will people make the effort to go through all the registries ( in addition to all the menu items) to find if what he is developing does not already exist ? :p

File: src/components/font_size_editor/font_size_editor.ts (hunk 1 above)
Comment (laa-odoo) [reply src/components/font_size_editor/font_size_editor.ts:34]: Necessary so as not to have too much padding before the icons,

Comment (laa-odoo) [reply src/components/font_size_editor/font_size_editor.ts:34]: so, in a next PR

File: src/components/top_bar/top_bar.ts [hunk 10]
```diff
@@ -254,30 +164,32 @@ css/* scss */ `
...
-  redoTool = false;
-  paintFormatTool = false;
-  fillColor: string = "#ffffff";
-  textColor: string = "#000000";
   menus: MenuItem[] = [];
+  editMenuItems = editMenuItems;
+  formatMenuItems = formatMenuItems;
+  viewMenuItems = viewMenuItems;
+  formatNumberMenuItem = createMenuItem(formatNumberMenuItemSpec);
+  isFeatureContextMenu = false;
```
Comment (LucasLefevre) [inline src/components/top_bar/top_bar.ts:192]: what's this flag ?

File: src/components/top_bar/top_bar.xml [hunk 11]
```diff
@@ -401,5 +275,43 @@
...
     </div>
+    <Menu
+      t-if="state.menuState.isOpen"
+      position="state.menuState.position"
+      menuItems="state.menuState.menuItems"
+      onClose="() => this.closeMenus()"
+    />
   </t>
+
+  <!-- <div class="o-dropdown">
```
Comment (LucasLefevre) [inline src/components/top_bar/top_bar.xml:286]: commented code

**Review by LucasLefevre — COMMENTED** [review]
Nice nice, I like it :+1: 

Some suggestions and highlights where we could go further. Most of them, if not all, can be done later.

Some general comments:
- css in bootstrap as much as possible
- I'm not sure about the naming of `isSelected`
- personal opinion: I don't like the "check" icon design. Too big, too bold
- more thinking about naming: to me, "menu item" refers to en entry/line in a dropdown menu, just like we have now. `<MenuItemButton/>` is actually just a `<Button/>`. In the same ideas, since menu items (`MenuItemSpec`) are used for buttons, they aren't really menu items only anymore. They are actually just generic "actions" ? That can be used as menu items or just simple buttons.
I feel like `MenuItemSpec` (and all its values) and `MenuItemButton` should have other names.
What do you think ?

**Comment by hokolomopo** [comment]
> * personal opinion: I don't like the "check" icon design. Too big, too bold

I think it's aan objective opinion that the check icon is kinda really ugly 🙃 There's an exmple of check icon in the filter-menu, we should probably use the same for both

**Review by hokolomopo — COMMENTED** [review]
some more comments

File: src/registries/menu_items_registry.ts (hunk 6 above)
Comment (laa-odoo) [reply src/registries/menu_items_registry.ts:51]: ok to replace ```isSelected``` by ```isActive```

File: src/registries/menus/items/sheet_menu_items.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      sheetIdTo,
+    });
+    env.model.dispatch("ACTIVATE_SHEET", { sheetIdFrom, sheetIdTo });
+  },
+};
+
+export const renameSheetMenuItem = (args: { renameSheetCallback: () => void }): MenuItemSpec => {
+  return {
+    name: _lt("Rename"),
+    action: () => args.renameSheetCallback(),
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/registries/menus/items/sheet_menu_items.ts:48]: nop

File: src/registries/menus/topbar_menu_registry.ts [hunk 13]
```diff
@@ -1,593 +1,499 @@
...
     sequence: 93,
-    action: ACTIONS.DELETE_CELL_SHIFT_UP,
   })
   .addChild("edit_delete_cell_shift_left", ["edit"], {
-    name: _lt("Delete cell and shift left"),
+    ...deleteCellShiftLeftMenuItem,
     sequence: 97,
-    action: ACTIONS.DELETE_CELL_SHIFT_LEFT,
   })
+  // Todo:  Change the coresponding registry --> view
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/registries/menus/topbar_menu_registry.ts:169]: forgot to remove

File: src/registries/menus/items/data_menu_items.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import * as ACTIONS from "./../menu_items_actions";
+
+export const sortRangeMenuItem: MenuItemSpec = {
+  name: _lt("Sort range"),
+  isVisible: ACTIONS.IS_ONLY_ONE_RANGE,
+};
+
+export const sortAscendingenuItem: MenuItemSpec = {
+  name: _lt("Ascending (A ⟶ Z)"),
+  action: ACTIONS.SORT_CELLS_ASCENDING,
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/registries/menus/items/data_menu_items.ts:12]: for another task ;)

File: src/components/top_bar/top_bar.ts (hunk 10 above)
Comment (laa-odoo) [reply src/components/top_bar/top_bar.ts:189]: Not well named.
Here, we need to differentiate between a menu that opens in the toolbar (eg the format button) and a menu that opens in the topbar (file/edit/view/insert/...).
The difference is that the topbar menu must open automatically with the mouse hover (when at least one of these is open) while this is not the case with the toolbar menus

Comment (laa-odoo) [reply src/components/top_bar/top_bar.ts:189]: isntToolbarMenu will be better

File: src/registries/menus/items/sheet_menu_items.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/registries/menus/items/sheet_menu_items.ts]: my bad, no worries, it works

File: src/components/top_bar/top_bar.ts [hunk 15]
```diff
@@ -254,30 +164,32 @@ css/* scss */ `
...
-  undoTool = false;
-  redoTool = false;
-  paintFormatTool = false;
-  fillColor: string = "#ffffff";
-  textColor: string = "#000000";
   menus: MenuItem[] = [];
+  editMenuItems = editMenuItems;
+  formatMenuItems = formatMenuItems;
+  viewMenuItems = viewMenuItems;
+  formatNumberMenuItem = createMenuItem(formatNumberMenuItemSpec);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/components/top_bar/top_bar.ts]: why ?

File: src/components/icons/icons.xml [hunk 16]
```diff
@@ -533,4 +533,25 @@
...
+  <t t-name="o-spreadsheet-Icon.PERCENT" owl="1">
+    <div class="o-text-icon">%</div>
+  </t>
+  <t t-name="o-spreadsheet-Icon.DECRASE_DECIMAL" owl="1">
+    <div class="o-text-icon">.0</div>
+  </t>
+  <t t-name="o-spreadsheet-Icon.INCREASE_DECIMAL" owl="1">
+    <div class="o-text-icon">.00</div>
+  </t>
+  <t t-name="o-spreadsheet-Icon.NUMBER_FORMATS" owl="1">
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/components/icons/icons.xml:554]: This is text not SVG, the correct SVG will be added in a special commit with all svgs missing :)

**Comment by laa-odoo** [comment]
@LucasLefevre, look good to me

**Review by LucasLefevre — APPROVED** [review]
robodoo r+
zé partiiiiii

**Comment by LucasLefevre** [comment]
robodoo r+
zé re-partiiiiii but without conflicts

**Comment by LucasLefevre** [comment]
robodoo rebase-ff

**Comment by LucasLefevre** [comment]
robodoo r+

**Pull request closed without merging**